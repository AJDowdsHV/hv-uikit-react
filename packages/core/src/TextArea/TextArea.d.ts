import { StandardProps } from "@material-ui/core";
import { HvInputLabelsProp, HvInputProps } from "../Input";

export interface HvTextAreaLabelsProp extends HvInputLabelsProp {
  /**
   * Text before the current char counter.
   */
  startCount: string;
  /**
   * Text between the current char counter and max value.
   */
  middleCount: string;
  /**
   * Text after the max value.
   */
  endCount: string;
}

export interface HvTextAreaProps extends StandardProps<HvInputProps, HvTextAreaClassKey, "labels"> {
  /**
   * An Object containing the various text associated with the input.
   */
  labels?: HvTextAreaLabelsProp;
  /**
   * The number of rows of the text area
   */
  rows?: number;
  /**
   * If ´true´ the component is resizable.
   */
  resizable?: boolean;
  /**
   * Auto-scroll: automatically scroll to the end on value changes.
   * Will stop if the user scrolls up and resume if scrolled to the bottom.
   */
  autoScroll?: boolean;
  /**
   * If true it isn't possible to pass the `maxCharQuantity`
   */
  blockMax?: boolean;
  /**
   * Props passed to the char count.
   */
  countCharProps?: object;
  /**
   * Indicates that user input is required on the form element.
   */
  isRequired?: boolean;
}

export type HvTextAreaClassKey =
  | "root"
  | "input"
  | "resize"
  | "inputRoot"
  | "inputRootDisabled"
  | "inputRootFocused"
  | "defaultWith"
  | "characterCounter"
  | "inline"
  | "separator"
  | "maxCharacter"
  | "currentCounter"
  | "disabled"
  | "container";

export default function HvTextArea(props: HvTextAreaProps): JSX.Element | null;
