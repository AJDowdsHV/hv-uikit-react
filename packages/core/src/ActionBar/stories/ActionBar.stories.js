import React from "react";
import { makeStyles } from "@material-ui/core";
import { HvActionContainer, HvButton, HvDropDownMenu } from "../..";

export default {
  title: "Patterns/Action Bar/Action Container",
  parameters: {
    componentSubtitle: null,
    usage: "import { HvActionContainer } from '@hv/uikit-react-core/dist'"
  },
  component: HvActionContainer
};

export const Main = () => {
  const useStyles = makeStyles(theme => ({
    root: {
      margin: "auto",
      backgroundColor: theme.hv.palette.atmosphere.atmo1,
      width: "1000px"
    },
    flexBox: {
      display: "flex"
    },
    separator: {
      marginRight: theme.hv.spacing.sm
    }
  }));

  const classes = useStyles();

  return (
    <div className={classes.root}>
      <HvActionContainer id="ActionContainer">
        <div className={classes.flexBox}>
          <div className={classes.separator}>
            <HvButton category="ghost" onClick={() => console.log("Save action")}>
              Save
            </HvButton>
          </div>
          <div>
            <HvButton category="ghost" onClick={() => console.log("Cancel action")}>
              Cancel
            </HvButton>
          </div>
        </div>
        <div>
          <HvButton category="ghost" onClick={() => console.log("Help action")}>
            Help
          </HvButton>
        </div>
      </HvActionContainer>
    </div>
  );
};

export const DualAction = () => {
  const useStyles = makeStyles(theme => ({
    root: {
      margin: "auto",
      backgroundColor: theme.hv.palette.atmosphere.atmo1,
      width: "1000px"
    },
    flexBox: {
      display: "flex"
    },
    separator: {
      marginRight: theme.hv.spacing.sm
    }
  }));

  const classes = useStyles();

  return (
    <div className={classes.root}>
      <HvActionContainer id="ActionContainer">
        <div className={classes.flexBox}>
          <div className={classes.separator}>
            <HvButton category="ghost" onClick={() => console.log("Save action")}>
              Save
            </HvButton>
          </div>
          <div>
            <HvButton category="ghost" onClick={() => console.log("Cancel action")}>
              Cancel
            </HvButton>
          </div>
        </div>
      </HvActionContainer>
    </div>
  );
};

DualAction.story = {
  parameters: {
    docs: {
      storyDescription: "Showcasing the action bar pattern with only two actions."
    }
  }
};

export const VariedActionBar = () => {
  const useStyles = makeStyles(theme => ({
    root: {
      margin: "auto",
      backgroundColor: theme.hv.palette.atmosphere.atmo1,
      width: "1000px"
    },
    flexBox: {
      display: "flex"
    },
    separator: {
      marginRight: theme.hv.spacing.sm
    }
  }));

  const classes = useStyles();

  return (
    <div className={classes.root}>
      <HvActionContainer id="ActionContainer">
        <div className={classes.flexBox}>
          <div className={classes.separator}>
            <HvButton category="ghost" onClick={() => console.log("Save action")}>
              Save
            </HvButton>
          </div>
          <div>
            <HvDropDownMenu
              onClick={(e, item) => console.log(item.label)}
              dataList={[{ label: "Delete" }, { label: "Update" }]}
            />
          </div>
        </div>
        <div>
          <HvButton category="ghost" onClick={() => console.log("Help action")}>
            Help
          </HvButton>
        </div>
      </HvActionContainer>
    </div>
  );
};

VariedActionBar.story = {
  parameters: {
    docs: {
      storyDescription: "Showcasing the action bar pattern with only two actions."
    }
  }
};
