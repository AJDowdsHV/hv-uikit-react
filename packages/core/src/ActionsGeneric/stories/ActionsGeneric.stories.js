/* eslint-disable react/prop-types */
import React from "react";
import { Add, Delete, Lock, Preview } from "@hv/uikit-react-icons";
import HvActionsGeneric from "..";

export default {
  title: "Components/Actions Generic",
  parameters: {
    componentSubtitle: null,
    usage: "import { HvActionsGeneric } from '@hv/uikit-react-core/dist'",
    dsVersion: "3.2.0",
  },
  component: HvActionsGeneric,
  decorators: [
    (Story) => (
      <div
        style={{
          display: "flex",
          justifyContent: "center",
        }}
      >
        <Story />
      </div>
    ),
  ],
};

export const Main = () => (
  <HvActionsGeneric
    actions={[
      { id: "add", label: "Add", iconCallback: () => <Add /> },
      { id: "delete", label: "Delete", iconCallback: () => <Delete /> },
      { id: "lock", label: "Lock", iconCallback: () => <Lock /> },
      { id: "put", label: "Preview", iconCallback: () => <Preview /> },
    ]}
    maxVisibleActions={2}
  />
);
