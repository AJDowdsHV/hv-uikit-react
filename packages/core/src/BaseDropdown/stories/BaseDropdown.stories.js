import React, { useState } from "react";
import { MoreOptionsVertical } from "@hv/uikit-react-icons";
import { useTheme } from "@material-ui/core";
import HvBaseDropdown from "..";
import HvInput from "../../Input";
import HvButton from "../../Button";
import HvTypography from "../../Typography";
import { HvFormElement, HvInfoMessage, HvLabel } from "../../Forms";

export default {
  title: "Components/BaseDropdown",
  parameters: {
    componentSubtitle: null,
    usage: "import { HvBaseDropdown } from '@hv/uikit-react-core/dist'"
  },
  component: HvBaseDropdown
};

export const Main = () => <HvBaseDropdown placeholder="Placeholder..." />;

Main.story = {
  parameters: {
    v3: true,
    pa11y: {
      ignore: ["region"]
    }
  }
};

export const WithContent = () => {
  const theme = useTheme();
  return (
    <HvBaseDropdown
      placeholder={<HvTypography id="labelTypography">Placeholder</HvTypography>}
      aria-labelledby="labelTypography"
    >
      <div style={{ width: "300px", height: "300px", background: theme.palette.atmo2 }}>
        <HvInput id="input-dropdown" />
      </div>
    </HvBaseDropdown>
  );
};

WithContent.story = {
  parameters: {
    v3: true
  }
};

export const WithForms = () => (
  <HvFormElement id="withForms" disabled={false}>
    <HvLabel style={{ paddingBottom: "8px", display: "block" }} label="Label" />
    <HvBaseDropdown placeholder="Placeholder..." />
    <HvInfoMessage disableGutter style={{ paddingTop: "8px", display: "block" }}>
      The description.
    </HvInfoMessage>
  </HvFormElement>
);

WithForms.story = {
  parameters: {
    v3: true
  }
};

export const Controlled = () => {
  const [open, setOpen] = useState(false);

  return (
    <>
      <HvButton onClick={() => setOpen(!open)}>change</HvButton>
      <p />
      <HvBaseDropdown
        placeholder="PlaceHolder..."
        expanded={open}
        onToggle={(e, s) => setOpen(s)}
      />
    </>
  );
};

Controlled.story = {
  parameters: {
    v3: true
  }
};

export const Custom = () => {
  const [open, setOpen] = useState(false);

  const theme = useTheme();

  const setterOpen = () => setOpen(!open);

  return (
    <HvBaseDropdown
      expanded={open}
      onClickOutside={setterOpen}
      component={
        <HvButton
          overrideIconColors={false}
          onClick={setterOpen}
          category="icon"
          aria-label="Dropdown menu"
          style={{ backgroundColor: open ? theme.palette.atmo1 : undefined }}
        >
          <MoreOptionsVertical />
        </HvButton>
      }
    >
      <div style={{ width: "500px", height: "500px", background: theme.palette.atmo2 }} />
    </HvBaseDropdown>
  );
};

Custom.story = {
  parameters: {
    v3: true
  }
};
