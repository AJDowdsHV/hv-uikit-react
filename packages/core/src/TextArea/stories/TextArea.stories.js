import React, { useState } from "react";
import { HvTextArea, HvButton, HvInput } from "../..";

export default {
  title: "Components/Text Area",
  parameters: {
    componentSubtitle: null,
    usage: "import { HvTextArea } from '@hv/uikit-react-core/dist'"
  },
  component: HvTextArea
};

export const Main = () => {
  const labels = {
    inputLabel: "Label",
    placeholder: "Enter value"
  };

  return <HvTextArea label="Text Area" labels={labels} id="test" width={610} />;
};

export const Resizable = () => {
  const labels = {
    inputLabel: "Label",
    placeholder: "Enter value"
  };

  return <HvTextArea label="Text Area" labels={labels} id="test" rows={5} resizable />;
};

Resizable.story = {
  parameters: {
    docs: {
      storyDescription: "Text area that allow resizing."
    }
  }
};

export const Limited = () => {
  const labels = {
    inputLabel: "Label",
    placeholder: "Enter value"
  };

  return <HvTextArea rows={5} labels={labels} maxCharQuantity={10} />;
};

Limited.story = {
  parameters: {
    docs: {
      storyDescription:
        "Text area that limits the quantity of character that can be introduced in the text area."
    }
  }
};

export const Disabled = () => {
  const labels = {
    inputLabel: "Label",
    placeholder: "Enter value"
  };

  return <HvTextArea label="Text Area" rows={5} labels={labels} maxCharQuantity={1500} disabled />;
};

Disabled.story = {
  parameters: {
    docs: {
      storyDescription: "Text area that does not allows any interaction."
    }
  }
};

export const Controlled = () => {
  const labels = {
    inputLabel: "Label",
    placeholder: "Enter value"
  };

  const btnStyle = {
    width: "130px",
    height: "32px",
    margin: "0 10px 30px 0"
  };

  const [value, setValue] = useState();

  // to be possible to change the input value by user action
  const setterValue = newValue => {
    setValue(newValue);
    return newValue;
  };

  return (
    <>
      <HvButton style={btnStyle} onClick={() => setValue("First value")}>
        First value
      </HvButton>
      <HvButton style={btnStyle} onClick={() => setValue("Second value")}>
        Second value
      </HvButton>
      <HvButton style={btnStyle} onClick={() => setValue("Third value")}>
        Third value
      </HvButton>

      <HvTextArea
        initialValue="Initial State"
        value={value}
        rows={5}
        labels={labels}
        onChange={setterValue}
      />
    </>
  );
};

Controlled.story = {
  parameters: {
    docs: {
      storyDescription: "Text area value altered from an outside component."
    }
  }
};

export const ControlledLimited = () => {
  const labels = {
    inputLabel: "Label",
    placeholder: "Enter value"
  };

  const btnStyle = {
    width: "130px",
    height: "32px",
    margin: "0 10px 30px 0"
  };

  const inpStyle = {
    marginBottom: "30px"
  };

  const inputLabels = {
    inputLabel: "Limit"
  };

  const [value, setValue] = useState();
  const [maxChar, setMaxChar] = useState(10);

  // to be possible to change the input value by user action
  const setterValue = newValue => {
    setValue(newValue);
    return newValue;
  };

  return (
    <>
      <HvButton style={btnStyle} onClick={() => setValue("First value")}>
        First value
      </HvButton>
      <HvButton style={btnStyle} onClick={() => setValue("Second value")}>
        Second value
      </HvButton>
      <HvButton style={btnStyle} onClick={() => setValue("Third value")}>
        Third value
      </HvButton>

      <HvInput
        style={inpStyle}
        initialValue={maxChar.toString()}
        onChange={newValue => setMaxChar(Number(newValue))}
        labels={inputLabels}
      />

      <HvTextArea
        initialValue="Initial State"
        value={value}
        rows={5}
        labels={labels}
        onChange={setterValue}
        maxCharQuantity={maxChar}
      />
    </>
  );
};

ControlledLimited.story = {
  parameters: {
    docs: {
      storyDescription: "Text area value altered from an outside component."
    }
  }
};
