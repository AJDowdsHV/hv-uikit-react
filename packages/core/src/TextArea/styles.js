const styles = {
  root: {
    display: "inline-block",
    width: "100%",
  },
  resizable: {
    width: "auto",
  },
  disabled: {},
  invalid: {},

  baseInput: {},

  input: {},
  inputResizable: {
    resize: "auto",
    minWidth: "150px",
    maxWidth: "610px",
  },

  labelContainer: {},
  label: {
    display: "block",
    float: "left",
    paddingBottom: "6px",
  },
  description: {
    display: "block",
    float: "left",
  },
  characterCounter: {
    display: "block",
    float: "right",
    textAlign: "right",
    marginBottom: "6px",
  },
};

export default styles;
