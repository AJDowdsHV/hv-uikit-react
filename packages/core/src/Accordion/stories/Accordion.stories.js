import React, { useState, useMemo } from "react";
import { makeStyles } from "@mui/styles";
import {
  HvAccordion,
  HvListContainer,
  HvListItem,
  HvInput,
  HvButton,
  HvTable,
  HvTableHead,
  HvTableHeader,
  HvTableRow,
  HvTableBody,
  HvTableCell,
  useHvData,
  useHvTableSticky,
} from "../..";

export default {
  title: "Display/Accordion",
  parameters: {
    componentSubtitle: null,
    usage: "import { HvAccordion } from '@hitachivantara/uikit-react-core'",
  },

  component: HvAccordion,
};

export const Main = () => {
  const useStyles = makeStyles(() => ({
    accordionSpacing: {
      "& > li": {
        paddingLeft: 32,
      },
    },
    accordionContainer: {
      maxWidth: 300,
    },
  }));
  const classes = useStyles();
  return (
    <div className={classes.accordionContainer}>
      <HvAccordion id="item1" label="Analytics" headingLevel={3}>
        <HvListContainer className={classes.accordionSpacing} interactive condensed>
          <HvListItem>Views</HvListItem>
          <HvListItem>Parameters</HvListItem>
        </HvListContainer>
      </HvAccordion>
      <HvAccordion id="item2" label="System" headingLevel={3}>
        <HvListContainer className={classes.accordionSpacing} interactive condensed>
          <HvListItem>Settings</HvListItem>
          <HvListItem>Network</HvListItem>
        </HvListContainer>
      </HvAccordion>
      <HvAccordion id="item3" label="Data" headingLevel={3}>
        <HvListContainer className={classes.accordionSpacing} interactive condensed>
          <HvListItem>Storage</HvListItem>
          <HvListItem>Memory</HvListItem>
        </HvListContainer>
      </HvAccordion>
    </div>
  );
};

export const Disabled = () => {
  const useStyles = makeStyles(() => ({
    accordionSpacing: {
      "& > li": {
        paddingLeft: 32,
      },
    },
    accordionContainer: {
      maxWidth: 300,
    },
  }));
  const classes = useStyles();
  return (
    <div className={classes.accordionContainer}>
      <HvAccordion id="disabled-item1" label="Analytics" headingLevel={3} disabled>
        <HvListContainer className={classes.accordionSpacing} interactive condensed>
          <HvListItem>Views</HvListItem>
          <HvListItem>Parameters</HvListItem>
        </HvListContainer>
      </HvAccordion>
      <HvAccordion id="disabled-item2" label="System" headingLevel={3}>
        <HvListContainer className={classes.accordionSpacing} interactive condensed>
          <HvListItem>Settings</HvListItem>
          <HvListItem>Network</HvListItem>
        </HvListContainer>
      </HvAccordion>
      <HvAccordion id="disabled-item3" label="Data" headingLevel={3} disabled>
        <HvListContainer className={classes.accordionSpacing} interactive condensed>
          <HvListItem>Storage</HvListItem>
          <HvListItem>Memory</HvListItem>
        </HvListContainer>
      </HvAccordion>
    </div>
  );
};

Disabled.parameters = {
  docs: {
    description: { story: "A common accordion with disable items. " },
  },
};

export const Controlled = () => {
  const useStyles = makeStyles((theme) => ({
    accordionSpacing: {
      "& > li": {
        paddingLeft: 32,
      },
    },
    accordionContainer: {
      maxWidth: 300,
    },
    buttonContainer: {
      "& button": {
        marginRight: theme.hvSpacing("xs"),
        marginBottom: theme.hvSpacing("md"),
      },
    },
  }));
  const classes = useStyles();
  const [expandedState, setExpandedState] = useState({ analytics: true, system: true, data: true });
  const handleToggle = (key) => {
    const newValue = { ...expandedState };
    newValue[key] = !newValue[key];
    setExpandedState(newValue);
  };
  const handleAll = (option) => {
    setExpandedState({ analytics: option, system: option, data: option });
  };
  return (
    <>
      <div className={classes.buttonContainer}>
        <HvButton category="secondary" onClick={() => handleToggle("analytics")}>
          Toggle Analytics
        </HvButton>
        <HvButton category="secondary" onClick={() => handleToggle("system")}>
          Toggle System
        </HvButton>
        <HvButton category="secondary" onClick={() => handleToggle("data")}>
          Toggle Data
        </HvButton>
        <HvButton category="secondary" onClick={() => handleAll(false)}>
          Close all
        </HvButton>
        <HvButton category="secondary" onClick={() => handleAll(true)}>
          Expand all
        </HvButton>
      </div>
      <div className={classes.accordionContainer}>
        <HvAccordion
          id="controlled-item1"
          label="Analytics"
          onChange={() => handleToggle("analytics")}
          expanded={expandedState.analytics}
        >
          <HvListContainer className={classes.accordionSpacing} interactive condensed>
            <HvListItem>Views</HvListItem>
            <HvListItem>Parameters</HvListItem>
          </HvListContainer>
        </HvAccordion>
        <HvAccordion
          id="controlled-item2"
          label="System"
          onChange={() => handleToggle("system")}
          expanded={expandedState.system}
        >
          <HvListContainer className={classes.accordionSpacing} interactive condensed>
            <HvListItem>Settings</HvListItem>
            <HvListItem>Network</HvListItem>
          </HvListContainer>
        </HvAccordion>
        <HvAccordion
          id="controlled-item3"
          label="Data"
          onChange={() => handleToggle("data")}
          expanded={expandedState.data}
        >
          <HvListContainer className={classes.accordionSpacing} interactive condensed>
            <HvListItem>Storage</HvListItem>
            <HvListItem>Memory</HvListItem>
          </HvListContainer>
        </HvAccordion>
      </div>
    </>
  );
};

Controlled.parameters = {
  docs: {
    description: { story: "An accordion controlled externally forcing expansion." },
  },
};

export const Form = () => {
  const useStyles = makeStyles(() => ({
    accordionContainer: {
      maxWidth: 300,
    },
    formContainer: {
      padding: "0 32px",
      "& > div": {
        paddingTop: 17,
      },
    },
  }));
  const classes = useStyles();
  return (
    <div className={classes.accordionContainer}>
      <HvAccordion id="personal" label="Personal Information" headingLevel={3}>
        <div className={classes.formContainer}>
          <HvInput id="input-name" label="Name" placeholder="Insert first name" required />
          <HvInput id="input-email" label="Email" placeholder="Insert your email" required />
          <HvInput id="input-phone" label="Phone" placeholder="Insert your phone number" />
          <HvInput id="input-extension" label="Extension" placeholder="Insert phone extension" />
          <HvInput id="input-country" label="Country" placeholder="Insert country name" />
          <HvInput id="input-province" label="City/Province" placeholder="Insert province name" />
        </div>
      </HvAccordion>
      <HvAccordion id="billing" label="Billing Address" headingLevel={3}>
        <div className={classes.formContainer}>
          <HvInput id="input-address" label="Address 1" placeholder="Insert first name" />
          <HvInput id="input-address2" label="Address 2" placeholder="Insert address" />
          <HvInput id="input-city" label="City" placeholder="Insert city name" />
          <HvInput id="input-state" label="State" placeholder="Insert state" />
          <HvInput id="input-code" label="Zip Code" placeholder="Insert code" />
        </div>
      </HvAccordion>
      <HvAccordion id="shipping" label="Shipping Address" headingLevel={3}>
        <div className={classes.formContainer}>
          <HvInput id="input-address-bill" label="Address 1" placeholder="Insert first name" />
          <HvInput id="input-address2-bill" label="Address 2" placeholder="Insert address" />
          <HvInput id="input-city-bill" label="City" placeholder="Insert city name" />
          <HvInput id="input-state-bill" label="State" placeholder="Insert state" />
          <HvInput id="input-code-bill" label="Zip Code" placeholder="Insert code" />
        </div>
      </HvAccordion>
    </div>
  );
};

Form.parameters = {
  docs: {
    description: { story: "An accordion composition with forms." },
  },
};

export const Typography = () => {
  const useStyles = makeStyles(() => ({
    accordionContainer: {
      width: "100%",
    },
  }));
  const classes = useStyles();

  const getData = () => [
    {
      id: 1,
      title: "A new hope",
      director: "George Lucas",
      producer: "Gary Kurzt",
      release: "1977-05-25",
    },
    {
      id: 2,
      title: "Empires strike back",
      director: "Irvin Kershner",
      producer: "Gary Kurzt",
      release: "1980-05-17",
    },
    {
      id: 3,
      title: "Return of the jedi",
      director: "Richard Marquand",
      producer: "Howard Kazanjian",
      release: "1983-05-25",
    },
  ];

  const getColumns = () => [
    {
      Header: "Title",
      accessor: "title",
      sticky: "left",
      width: 150,
      style: { textTransform: "capitalize" },
    },
    {
      Header: "Director",
      accessor: "director",
    },
    {
      Header: "Producer",
      accessor: "producer",
    },
    {
      Header: "Release Date",
      accessor: "release",
    },
  ];

  const data = useMemo(() => getData(), []);
  const columns = useMemo(() => getColumns(), []);

  const { getTableProps, getTableHeadProps, getTableBodyProps, prepareRow, headerGroups, rows } =
    useHvData(
      {
        columns,
        data,
        stickyHeader: true,
      },
      useHvTableSticky
    );

  return (
    <div className={classes.accordionContainer}>
      <HvAccordion id="film" label="Films" headingLevel={2}>
        <HvTable id="accordion-table" {...getTableProps()}>
          <HvTableHead {...getTableHeadProps()}>
            {headerGroups.map((headerGroup) => (
              <HvTableRow {...headerGroup.getHeaderGroupProps()}>
                {headerGroup.headers.map((col) => (
                  <HvTableHeader {...col.getHeaderProps()}>{col.render("Header")}</HvTableHeader>
                ))}
              </HvTableRow>
            ))}
          </HvTableHead>
          <HvTableBody {...getTableBodyProps()}>
            {rows.map((row) => {
              prepareRow(row);

              return (
                <HvTableRow {...row.getRowProps()}>
                  {row.cells.map((cell) => (
                    <HvTableCell {...cell.getCellProps()}>{cell.render("Cell")}</HvTableCell>
                  ))}
                </HvTableRow>
              );
            })}
          </HvTableBody>
        </HvTable>
      </HvAccordion>
    </div>
  );
};

Typography.parameters = {
  docs: {
    description: { story: "An accordion with a different button typography." },
  },
};
