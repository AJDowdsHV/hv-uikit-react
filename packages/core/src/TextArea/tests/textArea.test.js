/* eslint-env jest */

import React from "react";
import { mount, shallow } from "enzyme";

import { HvProvider, HvTextArea } from "../..";
import { Main } from "../stories/TextArea.stories";

describe("TextArea", () => {
  let wrapper;

  beforeEach(async () => {
    wrapper = shallow(
      <HvProvider>
        <Main />
      </HvProvider>
    );
  });

  it("should render correctly", () => {
    expect(wrapper.find(HvTextArea)).toMatchSnapshot();
  });
});

describe("TextArea Component", () => {
  let wrapper;

  beforeEach(async () => {
    wrapper = shallow(
      <HvProvider>
        <HvTextArea rows={4} initialValue="test" onChange={() => {}} />
      </HvProvider>
    );
  });

  it("should render correctly if opened", () => {
    expect(wrapper.find(HvTextArea)).toMatchSnapshot();
  });

  it("should save the current value length on change", () => {
    const value = "value";
    const onChangeMock = jest.fn(() => value);
    const wrapperMount = mount(
      <HvProvider>
        <HvTextArea rows={4} initialValue="test" onChange={onChangeMock} maxCharQuantity={10} />
      </HvProvider>
    );
    const instance = wrapperMount.find("HvTextArea").instance();
    instance.onChangeHandler(null, value);
    expect(onChangeMock).toHaveBeenCalled();
    expect(instance.state.currentValueLength).toBe(5);
  });

  it("should limit the current value length on change", () => {
    const value = "value value value";
    const onChangeMock = jest.fn(() => value);
    const wrapperMount = mount(
      <HvProvider>
        <HvTextArea
          rows={4}
          initialValue="test"
          onChange={onChangeMock}
          maxCharQuantity={5}
          blockMax
        />
      </HvProvider>
    );
    const instance = wrapperMount.find("HvTextArea").instance();
    instance.onChangeHandler(null, value);
    expect(onChangeMock).toHaveBeenCalled();
    expect(instance.state.currentValueLength).toBe(5);
  });

  it("shouldn't limit the current value length on change", () => {
    const value = "value value value";
    const onChangeMock = jest.fn(() => value);
    const wrapperMount = mount(
      <HvProvider>
        <HvTextArea initialValue="test" onChange={onChangeMock} maxCharQuantity={5} />
      </HvProvider>
    );
    const instance = wrapperMount.find("HvTextArea").instance();
    instance.onChangeHandler(null, value);
    expect(onChangeMock).toHaveBeenCalled();
    expect(instance.state.currentValueLength).toBe(17);
  });

  it("should render the count label correctly", () => {
    const wrapperMount = mount(
      <HvProvider>
        <HvTextArea
          rows={4}
          initialValue="test"
          labels={{
            startCount: "Inserted",
            middleCount: "of",
            endCount: "available",
          }}
          maxCharQuantity={10}
        />
      </HvProvider>
    );
    const label = wrapperMount.find("HvTextArea").find("span");
    expect(label.at(0).text()).toBe("Inserted 4 of 10 available");
  });

  //--------------------------

  const getInputInstance = (defaultProps, value) => {
    wrapper = mount(
      React.createElement(
        (props) => (
          <HvProvider>
            <HvTextArea
              value={props.value}
              rows={4}
              initialValue={props.value}
              onChange={props.onChangeMock}
              maxCharQuantity={props.maxCharQuantity}
              blockMax={props.blockMax}
            />
          </HvProvider>
        ),
        defaultProps
      )
    );
    wrapper.setProps({ value });
    wrapper.update();
    const inputInstance = wrapper.find("HvTextArea").instance();
    return inputInstance;
  };

  it("should save the current value length on change of value", () => {
    const value = "four";
    const onChangeMock = jest.fn(() => value);
    const defaultProps = {
      initialValue: "example",
      onChange: onChangeMock,
    };
    const instance = getInputInstance(defaultProps, value);
    expect(instance.state.currentValueLength).toBe(4);
  });

  it("should limit the current value length on change of value", () => {
    const defaultProps = {
      initialValue: "four",
      maxCharQuantity: 5,
      blockMax: true,
    };
    const instance = getInputInstance(defaultProps, "onethousand");
    expect(instance.state.currentValueLength).toBe(5);
  });

  it("shouldn't limit the current value length on change of value", () => {
    const defaultProps = {
      initialValue: "four",
    };
    const instance = getInputInstance(defaultProps, "onethousand");
    expect(instance.state.currentValueLength).toBe(11);
  });

  it("should scroll down on update when autoScroll", () => {
    const wrapperMount = mount(
      <HvProvider>
        <HvTextArea classes={{}} value="test" autoScroll />
      </HvProvider>
    );
    const instance = wrapperMount.find("HvTextArea").instance();
    const inputElement = wrapperMount.find("textarea").getDOMNode();

    expect(instance.state.autoScrolling).toBe(true);
    expect(instance.textInputRef.current).toBeDefined();
    expect(inputElement.scrollTop).toBe(0);

    Object.defineProperty(inputElement, "scrollHeight", { value: 30 });
    Object.defineProperty(inputElement, "clientHeight", { value: 10 });

    wrapperMount.setProps({ value: "trigger update" });
    expect(inputElement.scrollTop).toBe(20);
  });

  it("should stop and resume autoScrolling on scroll", () => {
    const wrapperMount = mount(
      <HvProvider>
        <HvTextArea classes={{}} value="test" autoScroll />
      </HvProvider>
    );
    const instance = wrapperMount.find("HvTextArea").instance();
    const inputElement = wrapperMount.find("textarea").getDOMNode();
    expect(instance.state.autoScrolling).toBe(true);
    Object.defineProperty(inputElement, "scrollHeight", { value: 30 });
    Object.defineProperty(inputElement, "clientHeight", { value: 10 });

    inputElement.scrollTop = 10;
    inputElement.dispatchEvent(new Event("scroll"));
    expect(instance.state.autoScrolling).toBe(false);

    inputElement.scrollTop = 20;
    inputElement.dispatchEvent(new Event("scroll"));
    expect(instance.state.autoScrolling).toBe(true);
  });
});
