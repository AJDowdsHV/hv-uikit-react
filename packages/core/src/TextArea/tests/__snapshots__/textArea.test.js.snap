// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Component should render correctly if opened 1`] = `
<WithStyles(WithLabels)
  initialValue="test"
  onChange={[Function]}
  rows={4}
/>
`;

exports[`TextArea withStyles should render correctly 1`] = `<WithStyles(WithLabels) />`;
